### ISSUE TITLE
DnsNameResolver not refreshing after a cache hit

### ISSUE THREAD

**Issue by risenberg** [grpc/grpc-java#6530]
Hello.
I implemented a `NameResolver` based on `DnsNameResolver` from version 1.19.0 - it has the same caching logic, but calls a high-level API instead of a DNS server. I encountered an error in the following scenario:
 1. The first resolution returns a faulty result and caches it
 2. `refresh()` is called before cache ttl expires, and short-circuits via
```
                if (resolver.shutdown || !cacheRefreshRequired()) {
                    return;
                }
```
The relevant code in current version of DnsNameResolver looks like
```
  private void resolve() {
    if (resolving || shutdown || !cacheRefreshRequired()) {
      return;
    }
    resolving = true;
    executor.execute(new Resolve(listener));
  }
```
 3. `refresh()` is never called again, I am stuck with a faulty result.

As I understand the problem, it is because in step 2 the listener is not called, neither `onAddresses()` nor `onError()`, so the calling code is waiting for callback and cannot schedule a refresh.
So is it a bug in `DnsNameResolver` code or is my approach wrong? How do I ensure that re-resolution eventually happens, apart from getting rid of caching?

**Comment by ran-su** [comment]
The change there(https://github.com/grpc/grpc-java/pull/5364) is not related to bug fix or anything like it.
I not familiar with your implementation, but from your description it looks like your implementation can be improved(when to cache, cache eviction, etc). I recommend you to double check the logic in your resolver.

**Comment by risenberg** [comment]
I copied the caching logic from `DnsNameResolver` entirely, just changed the source from which the addresses come. It is not clear to me who is responsible for updating in case of original `DnsNameResolver`. The docs say:
> A NameResolver does not need to automatically re-resolve on failure. Instead, the Listener is responsible for eventually (after an appropriate backoff period) invoking refresh().

But what about periodic re-resolving in case of success? There is no such thing in `DnsNameResolver`. Is it because some other entity is responsible for this, by creating a new `NameResolver` when appropriate? Is it not done because it seems unnecessary for DNS? If I implement one myself, where can I look for an example and how updating the address set will affect existing connections? This is unclear from the docs.

Anyway, I think there is a bug in those caching lines in `DnsNameResolver`, which is not triggered only because the cache is never used - once `DnsNameResolver` resolves something successfully, it is never accessed again (from what i've seen in logs while testing).

**Comment by ran-su** [comment]
I just looked the related `DnsNameResolver` logic in master branch, it looks right. I still think once your code find out the resolved result is faulty it should do some thing about it. For `refresh()` to work, the code need to clear the faulty `cachedResolutionResults` value.

**Comment by risenberg** [comment]
But who is responsible for periodic updates? Should my own resolver schedule them and call `Listener::onAddresses`? If so, what is the point of `refresh()` method? Or is grpc code responsible for calling `refresh()`? If so, how often is it supposed to? Can I control this timing? I found that it performs retries with exponential backoff on resolution failure via `ManagedChannelImpl.DelayedNameResolverRefresh`, but nothing about refreshes on success, either in docs or in code.

**Comment by ran-su** [comment]
Does not refresh on success is intended behavior. 
May I know what part of the first result is wrong? If the address is wrong it should schedule re-resolve. Is the service config in the result faulty? Do you see any logs, if yes could you please share them to us so we can take a look?

**Comment by risenberg** [comment]
> Does not refresh on success is intended behavior.
Then what is the point of having a cache? It will never be accessed.

> If the address is wrong it should schedule re-resolve.
You are right, I should not cache a result that I know to be faulty by itself. It will fix the problem I started with, but I have a broader concern. As I said, I retrieve the set of backend addresses via an API call instead of a DNS lookup. What if the set of backends changes afterwards? How do I notify grpc code about that? Looks like I have to do periodic updates in my resolver anyway.

**Comment by ran-su** [comment]
For your problem, adding those logic into your customized name resolver is the solution.

**Comment by risenberg** [comment]
I am still confused about caching, why it is implemented the way it is. Should I bother with it in my implementation? I mean, on error nothing is cached and after a successful resolve a refresh() is never called. So is there a scenario where the cached value is read?

**Comment by ran-su** [comment]
It is totally up to you to decide if you want caching in you application. As we do not know your business needs we can not give recommendations on your implementation. The `DnsNameResolver` was designed to handle our requirement for DNS NOT a template for all name resolvers. From the code, the `cachedResolutionResults` is not used again other than to decide if it needs a refresh.

**Comment by risenberg** [comment]
Thanks for clarifying that part. I still have another question - when and how is re-resolving triggered by grpc code? Does it happen if a request hits deadline? Or if a channel connection times out? Basically I am trying to do some client-side load balancing between my backends with a NameResolver. While I do not need a sophisticated algorithm, I still need to switch to a new backend if the old one becomes unresponsive. How does grpc find out that it must switch? Or should I use something other than a NameResolver for this task?

**Comment by creamsoup** [comment]
the interaction between DnsNameResolver and channel happens in the `ManagedChannelImpl` via `Listener` see [link](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L1304).

we recently (only on master and will be available after 1.27.0 - scheduled next week) implemented a feature that handles service config error. before this change, we has/assume api contract that the service config is valid. unfortunately, the logic is quite complex.

**Comment by creamsoup** [comment]
closing it due to inactivity, please feel free to reopen
